### PR TITLE
Dynamic analysis of the Trappist-1 system

### PR DESCRIPTION
**Pull request by pleroy**

### CODE REVIEW

**Comment by pleroy** [comment]
retest this please

File: astronomy/trappist_dynamics_test.cpp [hunk 1]
```diff
@@ -0,0 +1,1177 @@
...
+  constexpr static int count = 4;
+  Time period;
+  double x{};
+  double y{};
+  Time time_to_first_transit;
+};
+
+using SystemParameters = std::array<PlanetParameters, 7>;
+using Population = std::vector<SystemParameters>;
+using ComputeLogPdf =
```
Comment (eggrobin) [inline astronomy/trappist_dynamics_test.cpp]: ComputeLogUnnormalizedPosteriorPDF (and log_unnormalized_posterior_pdf below).

File: astronomy/trappist_gravity_model.proto.txt [hunk 2]
```diff
@@ -0,0 +1,39 @@
+gravity_model {
+  solar_system_frame : TRAPPIST
+
+  # Star.
+  body {
+    name : "Trappist-1A"
```
[the remark this answers is not in the document]
Comment (pleroy) [reply astronomy/trappist_gravity_model.proto.txt]: Done.

File: astronomy/trappist_dynamics_test.cpp [hunk 3]
```diff
@@ -0,0 +1,1172 @@
...
+    element.hyperbolic_excess_velocity = std::nullopt;
+    element.semiminor_axis = std::nullopt;
+    element.impact_parameter = std::nullopt;
+    element.semilatus_rectum = std::nullopt;
+    element.specific_angular_momentum = std::nullopt;
+    element.periapsis_distance = std::nullopt;
+    element.apoapsis_distance = std::nullopt;
+    element.longitude_of_periapsis = std::nullopt;
+    element.true_anomaly = std::nullopt;
+    element.hyperbolic_mean_anomaly = std::nullopt;
```
[the remark this answers is not in the document]
Comment (pleroy) [reply astronomy/trappist_dynamics_test.cpp]: Done, not sure if it's better.

File: astronomy/trappist_dynamics_test.cpp [hunk 4]
```diff
@@ -0,0 +1,1172 @@
...
+std::vector<KeplerianElements<Trappist>> const& Genome::elements() const {
+  return elements_;
+}
+
+void Genome::Mutate(std::mt19937_64& engine,
+                    int const generation) {
+  std::student_t_distribution<> distribution(1);
+
+  // The standard deviation of the distribution has a strong effect on the
+  // convergence of the algorithm: if it's too small we do not explore the
```
[the remark this answers is not in the document]
Comment (pleroy) [reply astronomy/trappist_dynamics_test.cpp]: Done.

File: astronomy/trappist_dynamics_test.cpp [hunk 5]
```diff
@@ -0,0 +1,1172 @@
...
+                 << (*genome.elements()[j].period -
+                     *best_genome_->elements()[j].period) / Second
+                 << " s";
+    }
+    LOG(ERROR) << "new T = " << *genome.elements()[j].period / Day << " d";
+  }
+}
+}  // namespace genetics
+
+// DEMCMC stands for Differential Evolution - Markov Chain Monte-Carlo.
```
[the remark this answers is not in the document]
Comment (pleroy) [reply astronomy/trappist_dynamics_test.cpp]: Done.

File: astronomy/trappist_dynamics_test.cpp [hunk 6]
```diff
@@ -0,0 +1,1172 @@
...
+// See for instance "A Markov Chain Monte Carlo version of the genetic algorithm
+// Differential Evolution: easy Bayesian computing for real parameter spaces",
+// Braak, 2006.
+namespace demcmc {
```
[the remark this answers is not in the document]
Comment (pleroy) [reply astronomy/trappist_dynamics_test.cpp]: Let's keep the name short, we use it several times at the end of this file.

File: astronomy/trappist_dynamics_test.cpp [hunk 7]
```diff
@@ -0,0 +1,1172 @@
...
+  elements.hyperbolic_excess_velocity = std::nullopt;
+  elements.semiminor_axis = std::nullopt;
+  elements.impact_parameter = std::nullopt;
+  elements.semilatus_rectum = std::nullopt;
+  elements.specific_angular_momentum = std::nullopt;
+  elements.periapsis_distance = std::nullopt;
+  elements.apoapsis_distance = std::nullopt;
+  elements.longitude_of_periapsis = std::nullopt;
+  elements.true_anomaly = std::nullopt;
+  elements.hyperbolic_mean_anomaly = std::nullopt;
```
[the remark this answers is not in the document]
Comment (pleroy) [reply astronomy/trappist_dynamics_test.cpp]: Done.

File: astronomy/trappist_dynamics_test.cpp [hunk 8]
```diff
@@ -0,0 +1,1172 @@
...
+  result.x = *elements.eccentricity * Cos(*elements.argument_of_periapsis);
+  result.y = *elements.eccentricity * Sin(*elements.argument_of_periapsis);
+  result.time_to_first_transit =
+      *elements.period / (2 * π * Radian) *
+      (π / 2 * Radian - *elements.argument_of_periapsis -
+       *elements.mean_anomaly);
+  return result;
+}
+
+std::vector<double> EvaluatePopulation(
```
[the remark this answers is not in the document]
Comment (pleroy) [reply astronomy/trappist_dynamics_test.cpp]: I started to do that and gave up.  Not in this PR at least.

File: astronomy/trappist_initial_state_jd_2457000_000000000.proto.txt [hunk 9]
```diff
@@ -1,90 +1,96 @@
+# Data from  "The nature of the TRAPPIST-1 exoplanets", Grimm et al., 2018.
+# The longitude of the ascending node are not known, but their distribution is
+# described in "Planet-planet occultations in TRAPPIST-1 and other exoplanet
+# systems", Luger et al., 2017, figure 7.  We take it from a normal distribution
+# of mean 0 deg and standard deviation 0.05 deg, as that appears to be the mode
+# in the figure.
```
Comment (eggrobin) [inline astronomy/trappist_initial_state_jd_2457000_000000000.proto.txt]: Thank you @pleroy, but our inclinations are in another paper!

Comment (pleroy) [reply astronomy/trappist_initial_state_jd_2457000_000000000.proto.txt]: And the periods.

**Pull request merged**